contents/algorithms/graph.md

```diff
@@ -27,29 +27,78 @@ A tree-like diagram could very well be a graph that allows for cycles and a naiv
 - **Uncommon** - Topological Sort, Dijkstra's algorithm
 - **Rare** - Bellman-Ford algorithm, Floyd-Warshall algorithm, Prim's algorithm, Kruskal's algorithm
 
-In coding interviews, graphs are commonly represented as 2-D matrices where cells are the nodes and each cell can traverse to its adjacent cells (up/down/left/right). Hence it is important that you be familiar with traversing a 2-D matrix. When recursively traversing the matrix, always ensure that your next position is within the boundary of the matrix. More tips for doing depth-first searches on a matrix can be found [here](https://discuss.leetcode.com/topic/66065/python-dfs-bests-85-tips-for-all-dfs-in-matrix-question/). A simple template for doing depth-first searches on a matrix goes like this:
+In coding interviews, graphs are commonly represented as 2-D matrices where cells are the nodes and each cell can traverse to its adjacent cells (up/down/left/right). Hence it is important that you be familiar with traversing a 2-D matrix. When traversing the matrix, always ensure that your current position is within the boundary of the matrix and has not been visited before. 
+
+A simple template for doing depth-first searches on a matrix goes like this:
 
 ```py
-def traverse(matrix):
+def dfs(matrix):
+  # check for an empty graph
+  if not matrix:
+    return []
   rows, cols = len(matrix), len(matrix[0])
   visited = set()
   directions = ((0, 1), (0, -1), (1, 0), (-1, 0))
-  def dfs(i, j):
-    if (i, j) in visited:
+
+  # Uses short circuiting to check whether current position is within the
+  # boundary and has not been visited before checking if it is valid
+  def pass_all_conditions(i, j):
+    return i in range(rows) and j in range(cols) and (i, j) not in visited
```

Comment:
`i in range(rows)` -> this is an O(N) operation and bloats your time complexity by an entire order.

Reply:
That's a good point. I have managed to pass using `range` checks on LeetCode with Python 3, hence I assumed that Python does this smartly.